### PR TITLE
Update to current Chef community conventions

### PR DESCRIPTION
**Pull request by jeffbyrnes**
~Lots of cleanup here, primarily driving towards Chef 17 compatibility.~

Provides:

* Updated support/config files, consistent w/ current Chef community conventions
* Unit tests are performed against latest Chef
* Integration tests are performed against multiple versions of Chef
* Linting is done via `cookstyle`

<details><summary><h3>[Update 2022-08-25]</h3></summary>

Buried in a [comment way down is this, from me:](https://github.com/DataDog/chef-datadog/pull/820#issuecomment-1184835657)

> I looked at the unit test output for https://github.com/DataDog/chef-datadog/pull/852, and it may be that this particular unit test didn’t get run? Quite a few [resources were untouched by that ChefSpec run](https://app.circleci.com/pipelines/github/DataDog/chef-datadog/597/workflows/11a235ec-06a2-433f-9576-e168755d749d/jobs/7625/steps?invite=true#step-107-82).
>
> I ran the unit tests locally against https://github.com/DataDog/chef-datadog/commit/e77a6c3e8cf203ffdabafb8167932d948ceafe74 & a lot of the tests failed so… I think maybe the nature of how the tests are run on main is obfuscating things.
>
> This PR, in part, fixes that & clears up the running of tests, and also ensures the integration tests run against multiple distros, distro versions, Chef versions, and Datadog Agent versions, which the current main branch does not do (despite appearances).

Something to note that this PR brings to bear. I’ve not attempted to _fix_ these test failures, b/c I think that’s out-of-scope for these changes, but we should be mindful all the same.
</details>

<details><summary><h3>[Update 2022-09-13]</h3></summary>

Figured out a bit more about the unit tests while doing a larger cleanup of them (a separate PR will appear for that once this one actually merges).

Check the individual commits to see what sorted that out in the end.
</details>

### [Update 2022-10-12]

* Rebased against `main` branch again
* Added `cookstyle --autofix` lint fixes to make builds totally green

### CODE REVIEW

**Comment by albertvaka** [comment]
Hi @jeffbyrnes! In my tests, the cookbook seems to work fine on Chef 17 without any changes. Did you find it was not the case for you? If there's any change that is needed, we should find a way to incorporate it while still keeping compatibility with older versions of Chef, though.

**Comment by jeffbyrnes** [comment]
@albertvaka Chef 17 wants resources to set `unified_mode true` for compatibility, and that also means using yum >= 7.0.0 (b/c that’s the first version that’s Chef 17-compatible).

As for older versions of Chef, only Chef 16 & Chef 17 are supported right now, Chef <= 15 are end-of-life. Also, testing across a matrix of that many Chef version is challenging, since CircleCI has limits on matrix size, and GitHub Actions charges based on action usage.

Relatedly, the test config wasn’t testing this against current OS versions (e.g., CentOS 8, Debian 10, Ubuntu 20.04), so that seems like a good thing to add, too.

Still, it’s y’all’s cookbook! I was following the example of the community (mainly as exemplified by @sous-chefs) in terms of cookbooks.

I’ll spin up a new branch & add the bare minimum of changes to try things out on Chef 17, and we can see what that looks like.

**Comment by jeffbyrnes** [comment]
So just to see what’s different between our testing, I took a look through the log from the [CircleCI Kitchen test from your commit adding Chef 17 testing](https://app.circleci.com/pipelines/github/DataDog/chef-datadog/442/workflows/e824873d-9393-4502-b61f-07590fed5ef4/jobs/4465), in 7b39ef2 and, uh, it seems as though it’s passing b/c it’s not actually testing against Chef 17. Or 16. Or any version except Chef 14:

```
🕙 16:35:21 jeffbyrnes in datadog on  bare-minimum-chef-17 [?] via  v3.0.0 
♠ grep 'packages.chef.io' ~/Downloads/datadog-log.txt 
       url	https://packages.chef.io/files/stable/chef/14.12.9/ubuntu/14.04/chef_14.12.9-1_amd64.deb
       downloading https://packages.chef.io/files/stable/chef/14.12.9/ubuntu/14.04/chef_14.12.9-1_amd64.deb
       url	https://packages.chef.io/files/stable/chef/14.12.9/el/6/chef-14.12.9-1.el6.x86_64.rpm
       downloading https://packages.chef.io/files/stable/chef/14.12.9/el/6/chef-14.12.9-1.el6.x86_64.rpm
       url	https://packages.chef.io/files/stable/chef/14.12.9/el/7/chef-14.12.9-1.el7.x86_64.rpm
       downloading https://packages.chef.io/files/stable/chef/14.12.9/el/7/chef-14.12.9-1.el7.x86_64.rpm
       url	https://packages.chef.io/files/stable/chef/14.12.9/debian/8/chef_14.12.9-1_amd64.deb
       downloading https://packages.chef.io/files/stable/chef/14.12.9/debian/8/chef_14.12.9-1_amd64.deb
       url	https://packages.chef.io/files/stable/chef/14.12.9/el/6/chef-14.12.9-1.el6.x86_64.rpm
       downloading https://packages.chef.io/files/stable/chef/14.12.9/el/6/chef-14.12.9-1.el6.x86_64.rpm
       url	https://packages.chef.io/files/stable/chef/14.12.9/el/7/chef-14.12.9-1.el7.x86_64.rpm
       downloading https://packages.chef.io/files/stable/chef/14.12.9/el/7/chef-14.12.9-1.el7.x86_64.rpm
       url	https://packages.chef.io/files/stable/chef/14.12.9/el/6/chef-14.12.9-1.el6.x86_64.rpm
       downloading https://packages.chef.io/files/stable/chef/14.12.9/el/6/chef-14.12.9-1.el6.x86_64.rpm
       url	https://packages.chef.io/files/stable/chef/14.12.9/el/7/chef-14.12.9-1.el7.x86_64.rpm
       downloading https://packages.chef.io/files/stable/chef/14.12.9/el/7/chef-14.12.9-1.el7.x86_64.rpm
```

So… this PR may be revealing that this cookbook isn’t being tested the way we _think_ it’s being tested, and showing us that things are in fact broken in ways that were previously hidden.

I’ve rebased my branch so it can merge again, and dropped the GitHub Actions bits b/c y’all like to use CircleCI. That said, it’s easier to see what’s breaking via GH Actions, b/c they allow for a larger matrix of tests. You can check that out [in my fork](https://github.com/jeffbyrnes/chef-datadog/actions/runs/1242482202). In many cases, it appears to be b/c a particular package is missing for a specific distro version.

I’ll see if I can figure out why the unit tests aren’t working on CircleCI, but for me locally:

```
Finished in 38.24 seconds (files took 2.53 seconds to load)
815 examples, 0 failures

Randomized with seed 34741

[Coveralls] Outside the CI environment, not sending data.
```

**Comment by jeffbyrnes** [comment]
Figured out the unit tests: they needed to have `CHEF_LICENSE: accept-no-persist` set in their environment.

**Comment by jeffbyrnes** [comment]
Hmm, looks like adjusting things to work on Chef 17 means dropping Chef 14 support:

```
[2021-09-21T20:54:07+00:00] FATAL: Chef::Exceptions::CookbookChefVersionMismatch: Cookbook 'yum' version '7.1.0' depends on chef version [">= 15.3"], but the running chef version is 14.15.6
```

**Comment by albertvaka** [comment]
I'm using kitchen to test, where we have different OSes and Chef versions. Since the `unified_mode` warnings are non-critical, I would rather keep compatibility with Chef 14 given the amount of people that still uses that version (even though it's EOLd).

**Comment by jeffbyrnes** [comment]
@albertvaka well, based on what CircleCI is doing with Kitchen, are you sure that it’s actually running against the newer versions of Chef? CircleCI is supposedly doing the same thing you are locally, but it’s doing all of it using Chef 14, which makes me wonder.

**Comment by albertvaka** [comment]
Circle doesn't run all the available tests.

**Comment by jeffbyrnes** [comment]
@albertvaka ah! With the changes I made in 3f84d20fa4b6aa93f6022a2f592ec97e4144e62e & f769ecf9fa59171f396c6816f2d8a03129abacab, now it does! However, to fit them into the limits that CircleCI places, the jobs break down into `integration-${agent_major_version}-${chef_version}-${platform}-${platform_version}`

This means that each job runs _every_ suite, meaning if any _one_ suite fails, the entire job fails.

The actual list of failing suites:

* datadog-integrations, amazonlinux-2
* datadog-integrations, centos-7
* datadog-integrations, centos-8
* datadog-integrations, debian-10
* datadog-integrations, debian-9
* datadog-integrations, ubuntu-1804
* datadog-integrations, ubuntu-2004
* dd-agent-iot, amazonlinux-2
* dd-agent-iot, centos-7
* dd-agent-iot, centos-8
* dd-agent-iot, debian-10
* dd-agent-iot, debian-9
* dd-agent-iot, ubuntu-1804
* dd-agent-iot, ubuntu-2004
* dd-integration-resource, amazonlinux-2
* dd-integration-resource, centos-7
* dd-integration-resource, centos-8
* dd-integration-resource, debian-10
* dd-integration-resource, debian-9
* dd-integration-resource, ubuntu-1804
* dd-integration-resource, ubuntu-2004
* system-probe, debian-10
* system-probe, ubuntu-1804
* system-probe, ubuntu-2004

You can see that `datadog-integrations`, `dd-integration-resource`, & `dd-agent-iot` break on every run, the error there is that the package is missing. The `system-probe` is more limited in terms of which platform + version combos it fails on.

**Comment by jeffbyrnes** [comment]
@albertvaka I’d forgotten that the main reason I went ahead & refactored the Kitchen config was that I was unable to run the tests as they are configured on the `main` branch, instead I get this error

```
♠ chef exec kitchen list
$$$$$$ Deprecated configuration detected:
require_chef_omnibus
Run 'kitchen doctor' for details.

/opt/chef-workstation/embedded/lib/ruby/gems/3.0.0/gems/chef-config-17.4.38/lib/chef-config/workstation_config_loader.rb:122:in `pwd': Too many open files - getcwd (Errno::EMFILE)
	from /opt/chef-workstation/embedded/lib/ruby/gems/3.0.0/gems/chef-config-17.4.38/lib/chef-config/workstation_config_loader.rb:122:in `locate_local_config'
	from /opt/chef-workstation/embedded/lib/ruby/gems/3.0.0/gems/chef-config-17.4.38/lib/chef-config/workstation_config_loader.rb:54:in `config_location'
	from /opt/chef-workstation/embedded/lib/ruby/gems/3.0.0/gems/chef-config-17.4.38/lib/chef-config/workstation_config_loader.rb:76:in `load'
	from /opt/chef-workstation/embedded/lib/ruby/gems/3.0.0/gems/test-kitchen-3.0.0/lib/kitchen/provisioner/chef_base.rb:239:in `initialize'
[snip]
```

If I reconfigure it so that setting the Chef version is injected via Env Var, like I did on this branch, then I’m able to run them. I’ll post the results once it finishes them all.

**Comment by jeffbyrnes** [comment]
Looking at [the most recent, as of this writing, CircleCI Kitchen tests](https://app.circleci.com/pipelines/github/DataDog/chef-datadog/527/workflows/41766576-c9dd-4e99-9d91-11c5d0924042/jobs/6878), we can see that integration tests are done on a _very_ limited set of combos of Chef versions, Datadog Agent versions, and OS distros/versions (extra line breaks added for readability):

```plain
♠ grep '\(Creating <\|Installing Chef\)' ~/Downloads/datadog-circleci.txt
-----> Creating <dd-agent-handler5-ubuntu-1404>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler5-centos-66>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler5-centos-77>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler5-debian-811>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler5-rocky-8>...
-----> Installing Chef 16.17.4 package

-----> Creating <dd-agent-handler6-centos-66>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler6-centos-77>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler6-rocky-8>...
-----> Installing Chef 16.17.4 package

-----> Creating <dd-agent-handler7-centos-66>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler7-centos-77>...
-----> Installing Chef 14.12 package

-----> Creating <dd-agent-handler7-rocky-8>...
-----> Installing Chef 16.17.4 package
```

I’ll narrow this down to only test those combinations, to create parity with what’s tested currently, rather than attempt to expand testing to additional combinations, as this PR currently does.

**Comment by jeffbyrnes** [comment]
And finally, all passing, with the exact same set of Kitchen suites that’s used on `main` right now.

**Comment by jeffbyrnes** [comment]
@albertvaka @jtappa refactored this significantly, updated the description to be relevant again.

This also provides fixes/changes that #838 #724 provide (albeit slightly differently).

**Review by jtappa — APPROVED** [review]
Markdown still looks good, 👍🏻 from docs. Thanks for the re-ping!

**Comment by jeffbyrnes** [comment]
@albertvaka absolutely! And yes, the style changes should be separate, to ensure any potential changes or regressions are clearly seen as caused by those vs. other adjustments.

I’ll try & spend some more time this week.

**Comment by jeffbyrnes** [comment]
@albertvaka ok, trimmed down this initial PR to a much smaller set of changes, focused on some reorg & cleanup, esp. of the unit tests.

**Comment by jeffbyrnes** [comment]
@albertvaka might be better to get #841 sorted first, since all the tests here aren’t going to work anymore due to the removal of things like `Gemfile`.

**Comment by jeffbyrnes** [comment]
@albertvaka using a `Gemfile` hasn’t been considered a good practice for Chef cookbooks for some time. ChefCo pivoted to Chef Workstation (née ChefDK) in 2014 as the supported & recommended way to develop cookbooks & library code.

I mentioned getting #841 approved & merged first b/c it takes care of updating the actual test elements to work without a `Gemfile`. It was a little tricky to break things up as much as I did, and it may make sense to collapse #841 back into this PR if you’d like to see green builds.

**Comment by jeffbyrnes** [comment]
Expanding on “why is a `Gemfile` not a thing Chef recommends anymore” it’s mostly to do w/ ensuring the version of Ruby, and the gems installed, will play nicely with Chef in a firmly dependable way. They vendor a copy of Ruby & the various gems that Chef needs as part of Chef Workstation, and this is as similar as possible to a production Chef Client install on a Chef Node.

Doing it with a `Gemfile` leaves one open to a lot more variation in your development environment vs. the production one (in terms of Ruby version, gems installed & gem versions installed, etc).

File: Berksfile
Comment (albertvaka) [inline Berksfile:6]: We need this for compatibility with older Chefs.

**Review by albertvaka — COMMENTED** [review]
Moving to Chef Workstation makes sense, so 👍 to removing the Gemfile, but can we continue to test old versions of Chef and OSes (eg: Chef 12 on Debian 8) using that?

File: metadata.rb
Comment (jeffbyrnes) [inline metadata.rb:37]: @albertvaka per your comments on the `Berksfile`, here are the two dependencies needed for older versions of Chef, expressed in a way that avoids duplication, and _also_ avoids them being included on newer versions of Chef (which can be problematic, esp. for Chef 17+).

File: Berksfile
Comment (jeffbyrnes) [reply Berksfile:6]: Please see https://github.com/DataDog/chef-datadog/pull/820/files#r839644281

File: Berksfile
Comment (jeffbyrnes) [reply Berksfile:10]: Please see https://github.com/DataDog/chef-datadog/pull/820/files#r839644281

File: .rubocop.yml
Comment (jeffbyrnes) [reply .rubocop.yml:42]: Gotcha, I’ll bring this back.

**Comment by jeffbyrnes** [comment]
So yes, testing on older versions of Chef is absolutely still possible. Like I’ve mentioned, we may want to #841 merged in prior to this one, since it'll become more clear how this can continue to be tested against various versions of Chef.

**Comment by jeffbyrnes** [comment]
Decided to fold #841 into this as b243d8b, to try & get a better test outcome.

**Comment by albertvaka** [comment]
👍 Makes sense. I'm all for modernizing this setup (so thank you for that!) but I'm worried about losing coverage for older chefs that many people use...

**Comment by jeffbyrnes** [comment]
@albertvaka oh for sure! There’s no loss in coverage though, this is testing all the same older versions of Chef. It also allows for the set of tests run to be expanded further in a way that’s more consistent & simpler than the previous implementation.

**Comment by jeffbyrnes** [comment]
FYI the unit tests are failing b/c they're outdated:

```none
Fauxhai::Exception::InvalidPlatform:
  Could not find platform 'centos/6.9' on the local disk and an HTTP error was encountered when fetching from Github. A list of available platforms is available at https://github.com/chefspec/fauxhai/blob/master/PLATFORMS.md
```

I’ll tweak that so they work properly again.

File: spec/recipes/dd-agent_spec.rb
Comment (albertvaka) [inline spec/recipes/dd-agent_spec.rb:1011]: What about these removed tests for older platforms, or the change from Ubuntu 14->16? How can we keep these tested?

File: spec/recipes/repository_spec.rb
Comment (albertvaka) [inline spec/recipes/repository_spec.rb:276]: Same here

File: kitchen.docker.yml
Comment (jeffbyrnes) [reply kitchen.docker.yml:7]: Yes, the tests won’t pass if that’s set to `true`. The deprecations still show up, but as warnings instead of errors.

File: spec/recipes/dd-agent_spec.rb
Comment (jeffbyrnes) [reply spec/recipes/dd-agent_spec.rb]: These are only the unit tests and unfortunately Fauxhai doesn’t provide for older versions if you want to also test current versions.

However! The integration tests _can_ be run against older distro versions, and are in general a far more “real” set of tests.

Right now those are a smaller set of tests, but they can be expanded. You can look back at the history of this PR and its builds to see me wrestling with that, since quite a few things pass the unit tests, but fail when run in reality as an integration test.

Comment (albertvaka) [reply spec/recipes/dd-agent_spec.rb]: Thanks a lot for your explanation, that makes sense.

**Comment by jeffbyrnes** [comment]
@albertvaka the linter won’t pass until we do the other style changes that you asked be taken out of this PR, so if you’re all good here, it’d be great to merge this in & move on to those changes, and perhaps also expand what the integration tests do to cover more combinations of Chef version + Linux distro.

File: metadata.rb
Comment (axl89) [inline metadata.rb:25]: @jeffbyrnes shouldn't we also exclude `chef_handler` as it's in the Chef boneyard and raises Chef-33 deprecation error?

**Comment by jeffbyrnes** [comment]
@axl89 rebased this, resolved my conflicts, and went ahead & merged in your PR and set the merge commit to close #854. Let’s see if it passes tests!

**Comment by jeffbyrnes** [comment]
As a reminder, the linting is not expected to pass, as this PR purposefully leaves all of the ChefSpec corrections out to avoid conflating multiple kinds of far-reaching changes.

**Comment by jeffbyrnes** [comment]
A few more adjustments to get the unit tests to pass, and to update the docs.

**Comment by jeffbyrnes** [comment]
Interesting: #852 seems to be the cause of [the lone unit test failure right now](https://app.circleci.com/pipelines/github/DataDog/chef-datadog/603/workflows/9db209c4-1a7e-4ff3-9da4-ade93ee7e973/jobs/7700/tests#failed-test-0).

I reverted that commit locally, and the unit tests all pass:

```plain
♠ chef exec rspec
[snip]

Finished in 27.5 seconds (files took 2.2 seconds to load)
859 examples, 0 failures

[snip]
```

That test is asserting that on Windows, a reinstall will occur if an older version is request:

```ruby
context 'when the agent version requested is lesser than the installed version' do
  let(:requested_agent_version) do
    '6.17.0'
  end

  it 'returns true' do
    expect(described_class.must_reinstall?(node)).to be true
  end
end
```

So I think maybe a subtle regression was introduced? This PR doesn’t meaningfully change that library code, so it _seems_ unlikely (but not impossible) that the regression was introduced here.

**Comment by jeffbyrnes** [comment]
I looked at the unit test output for #852, and it may be that this particular unit test didn’t get run? Quite a few [resources were untouched by that ChefSpec run](https://app.circleci.com/pipelines/github/DataDog/chef-datadog/597/workflows/11a235ec-06a2-433f-9576-e168755d749d/jobs/7625/steps?invite=true#step-107-82).

I ran the unit tests locally against e77a6c3 & a _lot_ of the tests failed so… I think maybe the nature of how the tests are run on `main` is obfuscating things.

This PR, in part, fixes that & clears up the running of tests, and also ensures the integration tests run against multiple distros, distro versions, Chef versions, and Datadog Agent versions, which the current `main` branch does not do (despite appearances).

**Comment by axl89** [comment]
@jeffbyrnes be advised on [this comment](https://github.com/DataDog/chef-datadog/pull/854#issuecomment-1185251602) from the cookbook maintainers. Maybe we have to maintain Chef 12.7 or higher :disappointed:

**Comment by jeffbyrnes** [comment]
It's possible to accommodate; I'll tweak it to see about that.

**Comment by jeffbyrnes** [comment]
Something-something my accommodation didn’t quite work… I’ll take another stab at this soon.

**Comment by jeffbyrnes** [comment]
Haven’t forgotten about this, haven’t have much free time lately.

**Comment by jeffbyrnes** [comment]
Ok, we’re back to a [single unit test failure](https://app.circleci.com/pipelines/github/DataDog/chef-datadog/614/workflows/80932f21-90f7-4310-a33b-8b336309dc8a/jobs/7813?invite=true#step-103-44).

I think the Coveralls error is b/c these tests are running from my fork.

As for the linter errors, basically all of those are auto-correctable, with two exceptions:

```bash
Offenses:

recipes/dd-handler.rb:66:60: C: [Correctable] Style/CollectionCompact: Use compact instead of reject { |_, v| v.nil? }.
    extra_config = node['datadog']['handler_extra_config'].reject { |_, v| v.nil? }
                                                           ^^^^^^^^^^^^^^^^^^^^^^^^
spec/recipes/dd-agent_spec.rb:998:11: W: Chef/Deprecations/DeprecatedChefSpecPlatform: Use currently supported platforms in ChefSpec listed at https://github.com/chefspec/fauxhai/blob/main/PLATFORMS.md. Fauxhai / ChefSpec will perform fuzzy matching on platform version so it's always best to be less specific ie. 10 instead of 10.3 (https://docs.chef.io/workstation/cookstyle/chef_deprecations_deprecatedchefspecplatform)
          ChefSpec::SoloRunner.new( ...
          ^^^^^^^^^^^^^^^^^^^^^^^^^

260 files inspected, 2 offenses detected, 1 more offense can be corrected with `rubocop -A`
```

The first one can be corrected, but it’s not 100% safe.

The second is best handled via a `rubocop:disable` control comment, since it seems there’s desire to continue doing unit tests against deprecated OS versions.

I’ll refrain from committing & pushing up those linter fixes just yet, so this can be reviewed in its current state, and folks can chime in if the linter fixes should be included, or handled via a followup PR.

**Comment by jeffbyrnes** [comment]
And I tracked down that single unit test failure: turns out, `libraries/recipe_helpers` was being tested by _two_ sets of unit tests.

I deleted the older one in 12b321a, and now we’ve got all our units & integration tests passing here.

I can bring in the `cookstyle` changes to get the linter passing as well, but seems like this PR is already _plenty_ big enough.

@axl89, @albertvaka, @jtappa, this really, finally, is in a place to be merged in.

Happy to discuss further.

**Review by jtappa — APPROVED** [review]
Small suggestions/requests for change from docs in terms of grammar and translation processes, but otherwise LGTM and I am marking as approved so it's not blocked on me for re-review.

**Comment by AliDatadog** [comment]
Closed for inactivity.

**Pull request closed without merging**
